modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SemiJoinNode.java

```diff
@@ -0,0 +1,234 @@
+                /** {@inheritDoc} */
+                @Override public void push(Row row) {
+                    pushLeft(row);
+                }
+
+                /** {@inheritDoc} */
+                @Override public void end() {
+                    endLeft();
+                }
+
+                /** {@inheritDoc} */
+                @Override public void onError(Throwable e) {
+                    SemiJoinNode.this.onError(e);
+                }
+            };
+        else if (idx == 1)
+            return new Downstream<Row>() {
+                /** {@inheritDoc} */
+                @Override public void push(Row row) {
+                    pushRight(row);
+                }
+
+                /** {@inheritDoc} */
+                @Override public void end() {
+                    endRight();
+                }
+
+                /** {@inheritDoc} */
+                @Override public void onError(Throwable e) {
+                    SemiJoinNode.this.onError(e);
+                }
+            };
+
+        throw new IndexOutOfBoundsException();
+    }
+
+    /** */
+    private void pushLeft(Row row) {
+        checkThread();
+
+        assert downstream != null;
+        assert waitingLeft > 0;
+
+        waitingLeft--;
+
+        leftInBuf.add(row);
+
+        flushFromBuffer();
+    }
+
+    /** */
+    private void pushRight(Row row) {
+        checkThread();
+
+        assert downstream != null;
+        assert waitingRight > 0;
+
+        waitingRight--;
+
+        rightMaterialized.add(row);
+
+        if (waitingRight == 0)
+            sources.get(1).request(waitingRight = IN_BUFFER_SIZE);
+    }
+
+    /** */
+    private void endLeft() {
+        checkThread();
+
+        assert downstream != null;
+        assert waitingLeft > 0;
+
+        waitingLeft = -1;
+
+        flushFromBuffer();
+    }
+
+    /** */
+    private void endRight() {
+        checkThread();
+
+        assert downstream != null;
+        assert waitingRight > 0;
+
+        waitingRight = -1;
+
+        flushFromBuffer();
+    }
+
+    /** */
+    private void onError(Throwable e) {
+        checkThread();
+
+        assert downstream != null;
+
+        downstream.onError(e);
+    }
+
+    /** */
+    private void flushFromBuffer() {
+        inLoop = true;
+        try {
+            if (waitingRight == -1) {
+                while (requested > 0 && (left != null || !leftInBuf.isEmpty())) {
+                    if (left == null)
+                        left = leftInBuf.remove();
+
+                    boolean matched = false;
+
+                    while (!matched && requested > 0 && rightIdx < rightMaterialized.size()) {
+                        Row row = handler.concat(left, rightMaterialized.get(rightIdx++));
+
+                        if (!cond.test(row))
+                            continue;
+
+                        requested--;
+                        matched = true;
+                        downstream.push(left);
```

Comment:
If I understand the semantics of semi-join right, it should emit only one row for each match. For example [here](https://sqlperformance.com/2018/02/sql-plan/row-goals-part-2-semi-joins) I see:

> A semi join returns a row from one join input (A) if there is at least one matching row on the other join input (B).
> The essential differences between a semi join and a regular join are:
> - Semi join either returns each row from input A, or it does not. No row duplication can occur.
> - Regular join duplicates rows if there are multiple matches on the join predicate.
> - Semi join is defined to only return columns from input A.
> - Regular join may return columns from either (or both) join inputs.

So `rightMaterialized` can be a sort of set here.

Reply:
>  it should emit only one row for each match
>
And it does so. Please see test https://github.com/apache/ignite/pull/7915/files#diff-57b3dae5cc807015f62943792c5c70ebR343